engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -0,0 +1,508 @@
+import com.oracle.truffle.api.interop.UnsupportedTypeException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import com.oracle.truffle.api.profiles.LoopConditionProfile;
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.time.LocalTime;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.node.callable.ExecuteCallNode;
+import org.enso.interpreter.node.expression.builtin.meta.EqualsNodeGen.InvokeEqualsNodeGen;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.callable.UnresolvedConversion;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.callable.atom.Atom;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.callable.function.Function;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.number.EnsoBigInteger;
+import org.enso.interpreter.runtime.state.State;
+import org.enso.pkg.QualifiedName;
+import org.enso.polyglot.MethodNames;
+
+@BuiltinMethod(
+    type = "Any",
+    name = "equals_builtin",
+    description = "Implementation of Any.=="
+)
+@GenerateUncached
+@SpecializationStatistics.AlwaysEnabled
+public abstract class EqualsNode extends Node {
+
+  protected static String EQUALS_MEMBER_NAME = MethodNames.Function.EQUALS;
+
+  public static EqualsNode build() {
+    return EqualsNodeGen.create();
+  }
+  public abstract boolean execute(@AcceptsError Object self, @AcceptsError Object right);
+
+  /** Primitive values **/
+
+  @Specialization
+  boolean equalsLong(long self, long other) {
+    return self == other;
+  }
+
+  @Specialization
+  boolean equalsBoolean(boolean self, boolean other) {
+    return self == other;
+  }
+
+  @Specialization
+  boolean equalsDouble(double self, double other) {
+    return self == other;
+  }
+
+  @Specialization
+  @TruffleBoundary
+  boolean equalsBigInt(EnsoBigInteger self, EnsoBigInteger otherBigInt) {
+    return self.equals(otherBigInt);
+  }
+
+  /** Enso specific types **/
+
+  @Specialization
+  boolean equalsUnresolvedSymbols(UnresolvedSymbol self, UnresolvedSymbol otherSymbol,
+      @Cached EqualsNode equalsNode) {
+    return self.getName().equals(otherSymbol.getName())
+        && equalsNode.execute(self.getScope(), otherSymbol.getScope());
+  }
+
+  @Specialization
+  boolean equalsUnresolvedConversion(UnresolvedConversion selfConversion, UnresolvedConversion otherConversion,
+      @Cached EqualsNode equalsNode) {
+    return equalsNode.execute(selfConversion.getScope(), otherConversion.getScope());
+  }
+
+  /** Interop libraries **/
+
+  @Specialization(guards = {
```

Comment:
I don't see specialization for `Type`!? I am not sure how comparing of types work and whether it is tested somehere...

Btw. in the light of [getMetaObject implementation](https://www.pivotaltracker.com/story/show/183686167),  the _"type specialization"_ should precede this `isMetaObject` check. `type1 == type2` is certainly cheaper than `InteropLibrary` check.

Reply:
> I don't see specialization for Type!? I am not sure how comparing of types work and whether it is tested somehere...

Correct, it is not yet tested anywhere. I have added tests to `Equals_Spec.enso`, and I will add a specialization that will handle the types.

Reply:
This is blocked by https://www.pivotaltracker.com/story/show/184092284 - `MyType == MyType` dispatches to an instance method `Any.==` which is a current limitation. For now, I won't add a specialization for types. Moreover, type equality is not checked anywhere.